datadog/dogshell/service_level_objective.py

```diff
@@ -0,0 +1,418 @@
+# stdlib
+import argparse
+import json
+
+# 3p
+from datadog.util.format import pretty_json
+
+# datadog
+from datadog import api
+from datadog.dogshell.common import report_errors, report_warnings
+
+
+class MonitorClient(object):
+    @classmethod
+    def setup_parser(cls, subparsers):
+        parser = subparsers.add_parser(
+            "service_level_objective",
+            help="Create, edit, and delete service level objectives",
+        )
+
+        verb_parsers = parser.add_subparsers(title="Verbs", dest="verb")
+        verb_parsers.required = True
+
+        create_parser = verb_parsers.add_parser("create", help="Create a SLO")
+        create_parser.add_argument(
+            "--type",
+            required=True,
+            help="type of the SLO, e.g.",
+            choices=["metric", "monitor"],
+        )
+        create_parser.add_argument("--name", help="name of the SLO", default=None)
+        create_parser.add_argument(
+            "--description", help="description of the SLO", default=None
+        )
+        create_parser.add_argument(
+            "--tags", help="comma-separated list of tags", default=None
+        )
+        create_parser.add_argument(
+            "--thresholds",
+            help="comma separated list of <timeframe>:<target>[:<warning>]",
+            required=True,
+        )
+        create_parser.add_argument(
+            "--numerator",
+            help="numerator metric query (sum of good events)",
+            default=None,
+        )
+        create_parser.add_argument(
+            "--denominator",
+            help="denominator metric query (sum of total events)",
+            default=None,
+        )
+        create_parser.add_argument(
+            "--monitor_ids", help="explicit monitor_ids to use (CSV)", default=None
+        )
+        create_parser.add_argument(
+            "--monitor_search", help="monitor search terms to use", default=None
+        )
+        create_parser.set_defaults(func=cls._create)
+
+        file_create_parser = verb_parsers.add_parser(
+            "fcreate", help="Create a SLO from file"
+        )
+        file_create_parser.add_argument(
+            "file", help="json file holding all details", type=argparse.FileType("r")
+        )
+        file_create_parser.set_defaults(func=cls._file_create)
+
+        update_parser = verb_parsers.add_parser("update", help="Update existing SLO")
+        update_parser.add_argument(
+            "slo_id", help="SLO to replace with the new definition"
+        )
+        update_parser.add_argument(
+            "--type",
+            required=True,
+            help="type of the SLO (must specify it's original type)",
+            choices=["metric", "monitor"],
+        )
```

Comment:
Why is this required if it mustn't change ?

Reply:
this is required but the dogshell doesn't currently know what the current type is (though I could make it query and find out). this is more so it knows how to handle the update correctly.

Reply:
I see, let's leave like this then